File: 3-Tutorial/Praktikum 3/1/Rekening.cpp
// Nama 	: Naufal Dean Anugrah
// NIM		: 13518123
// Tanggal	: 6 Februari 2020
// Topik 	: Inheritance

#include <iostream>
#include "Rekening.h"

using namespace std;

// Konstruktor untuk menginisialisasi saldo
// Set saldo 0.0 apabila saldo bernilai negatif
Rekening::Rekening(double saldo) {
	this->saldo = (saldo < 0.0) ? (0.0) : (saldo);
}

// Getter, Setter
void Rekening::setSaldo(double saldo) {
	this->saldo = saldo;
}
double Rekening::getSaldo() const {
	return this->saldo;
}

// Member Function
// Tambahkan sejumlah nominal uang ke saldo
void Rekening::simpanUang(double deltaSaldo) {
	this->saldo += deltaSaldo;
}

// Kurangkan sejumlah nominal uang dari saldo apabila saldo mencukupi (tidak negatif setelah dikurangkan)
// Kembalikan boolean yang mengindikasikan apakah uang berhasil ditarik atau tidak
bool Rekening::tarikUang(double deltaSaldo) {
	if (this->saldo < deltaSaldo) {
		return false;
	} else {
		this->saldo -= deltaSaldo;
		return true;
	}
}

File: Java/7-Tutorial/3/Rekening.java
// Nama		: Naufal Dean Anugrah
// NIM		: 13518123
// Tanggal	: 12 Maret 2020
// Topik 	: Pengenalan java

abstract class Rekening {
	protected String nama;
	protected double saldo;
	protected double sukuBunga;

	public Rekening(String nama, double saldo, double sukuBunga) {
		this.nama = nama;
		this.sukuBunga = sukuBunga;
		if (saldo < 0) {
			this.saldo = 0;
		} else {
			this.saldo = saldo;
		}
	}

	abstract public void setor(double uang);
	abstract public void tarik(double uang);
	abstract public void update();

	public double hitungBiaya() {
		return Math.min(0.1 * this.saldo, 10.0);
	}

	public String getNama() {
		return this.nama;
	}

	public double getSaldo() {
		return this.saldo;
	}

	public double getSukuBunga() {
		return this.sukuBunga;
	}
}
